doc/learn/syntax-trees-with-typescript.md

```diff
@@ -0,0 +1,277 @@
+The core types are available in a types only package: [`@types/unist`](https://www.npmjs.com/package/@types/unist).
+The main type is `Node`.
+Everything else extends it.
+`Literal` and `Parent` are more practical types which also extend `Node`.
+
+The types provided by unist are abstract interfaces.
+In many cases, you will instead use more specific interfaces depending on what
+language you’re working with.
+Each language supported by unified, like markdown, HTML, and XML, has its own
+syntax tree standard which extends `unist`.
+
+Let’s take a look at these.
+
+### UNIST
+
+#### `Node`
+
+`Node` is the syntactic unit of syntax trees.
+Each node extends `Node` (sometimes indirectly through `Literal` or `Parent`)
+and set `type` to a [string literal](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types)
+The type field tells us what kind of syntax the node is.
+This field uniquely identifies a kind of content (in TypeScript parlance a
+[discriminated union](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions)).
+For example in Markdown (mdast) `Node` will be extended to make different kinds
+of content such as a `Heading` or `Link`, which respectively use a `type` field
+of `'heading'` and `'link'`.
+
+A node also can optionally include a `Data` interface at the `data` field.
+This is an object (dictionary) that stores extra metadata which is not standard
+to the node but defined by the ecosystem (utilities and plugins).
+
+When a syntax tree is parsed from a file, it includes positional information:
+a `Position` interface at the `position` field.
+This describes where the node occurred in the source file.
+
+```ts
+/**
+ * Syntactic units in unist syntax trees are called nodes.
+ */
+interface Node {
+  /**
+   * The variant of a node.
+   */
+  type: string
+
+  /**
+   * Information from the ecosystem.
+   */
+  data?: Data | undefined
+
+  /**
+   * Location of a node in a source document.
+   * Must not be present if a node is generated.
+   */
+  position?: Position | undefined
+}
+
+/**
+ * Information associated by the ecosystem with the node.
+ * Space is guaranteed to never be specified by unist or specifications
+ * implementing unist.
+ */
+export interface Data {
+  [key: string]: unknown
+}
+
+/**
+ * Location of a node in a source file.
+ */
+export interface Position {
+  /**
+   * Place of the first character of the parsed source region.
+   */
+  start: Point
+
+  /**
+   * Place of the first character after the parsed source region.
+   */
+  end: Point
+
+  /**
+   * Start column at each index (plus start line) in the source region,
+   * for elements that span multiple lines.
+   */
+  indent?: number[] | undefined
+}
+
+```
+
+#### `Literal`
+
+`Literal` extends `Node` and adds a `value` property.
+For example a markdown `code` node extends `Literal` and sets `value` to be a `string`.
```

Comment:
`'code'` or `Code` or `code`? I’d say use `` `Code` `` for TS types, `` `'code'` `` for the type string, and just the word code for the construct.

Reply:
updated